File: crabzilla-example1/crabzilla-example1-vertx/src/test/java/JacksonJsonTest.java
import com.fasterxml.jackson.databind.ObjectMapper;
import crabzilla.example1.aggregates.customer.CustomerId;
import crabzilla.example1.aggregates.customer.commands.CreateCustomerCmd;
import crabzilla.example1.aggregates.customer.events.CustomerActivated;
import crabzilla.example1.aggregates.customer.events.CustomerCreated;
import crabzilla.model.Command;
import crabzilla.model.UnitOfWork;
import crabzilla.model.Version;
import io.vertx.core.json.Json;
import lombok.val;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.time.Instant;
import java.util.Collections;
import java.util.UUID;

import static java.util.Arrays.asList;
import static org.assertj.core.api.Assertions.assertThat;

public class JacksonJsonTest {

  ObjectMapper mapper = Json.mapper;

  @BeforeEach
  public void setup() {
    mapper.findAndRegisterModules();
    mapper.enableDefaultTyping(ObjectMapper.DefaultTyping.NON_FINAL);
  }

  @Test
  public void one_event() throws Exception {

    val id = new CustomerId(UUID.randomUUID().toString());
    val command = new CreateCustomerCmd(UUID.randomUUID(), id, "customer1");
    val event = new CustomerCreated(id, command.getName());
    val uow1 = UnitOfWork.unitOfWork(command, Version.create(1), Collections.singletonList(event));

    val uowAsJson = mapper.writeValueAsString(uow1);

    System.out.println(mapper.writerFor(Command.class).writeValueAsString(command));

    System.out.println(uowAsJson);

    val uow2 = mapper.readValue(uowAsJson, UnitOfWork.class);

    assertThat(uow2).isEqualTo(uow1);

  }

  @Test
  public void more_events() throws Exception {

    val id = new CustomerId("customer#1");
    val command = new CreateCustomerCmd(UUID.randomUUID(), id, "customer1");
    val event1 = new CustomerCreated(id, command.getName());
    val event2 = new CustomerActivated("a rgood reason", Instant.now());

    val uow1 = UnitOfWork.unitOfWork(command, Version.create(1), asList(event1,  event2));

    val uowAsJson = mapper.writeValueAsString(uow1);

    //System.out.println(uowAsJson);

    val uow2 = mapper.readValue(uowAsJson, UnitOfWork.class);

    assertThat(uow2).isEqualTo(uow1);

  }

}

File: crabzilla-core/src/test/java/io/github/crabzilla/JacksonJsonTest.kt
package io.github.crabzilla

import io.vertx.core.Vertx
import org.junit.jupiter.api.BeforeEach

class JacksonJsonTest {

  lateinit var vertx: Vertx

  @BeforeEach
  fun setUp() {
    vertx = Vertx.vertx()
    initVertx(vertx)
  }

//  @Test
//  fun commandToJsonMustWork() {
//
//    val id = CustomerId(1)
//    val command = CreateCustomer(UUID.randomUUID(), id, "cust-1")
//
//    val cmdAsJson = commandToJson(command)
//
//    assertThat(Json.mapper.readValue(cmdAsJson, CreateCustomer::class.java)).isEqualTo(command)
//
//  }
//
//  @Test
//  fun commandFromJsonMustWork() {
//
//    val id = CustomerId(1)
//    val command = CreateCustomer(UUID.randomUUID(), id, "cust-1")
//
//    val cmdAsJson = commandToJson(command)
//    val commandFromJson = commandFromJson(cmdAsJson)
//
//    assertThat(command).isEqualTo(commandFromJson)
//
//  }
//
//  @Test
//  @Throws(Exception::class)
//  fun listOfEventsToJsonMustWork() {
//
//    val id = CustomerId(1)
//    val command = CreateCustomer(UUID.randomUUID(), id, "cust-1")
//
//    val event1 = CustomerCreated(id, command.name)
//    val event2 = CustomerActivated("a good reason", Instant.now())
//
//    val eventsList = asList(event1, event2)
//
//    val asJson = listOfEventsToJson(eventsList)
//
//    assertThat(Json.mapper.readerFor(eventsListType).readValue<List<DomainEvent>>(asJson)).isEqualTo(eventsList)
//
//    println(asJson)
//
//  }
//
//  @Test
//  fun listOfEventsFromJsonMustWork() {
//
//    val id = CustomerId(1)
//    val command = CreateCustomer(UUID.randomUUID(), id, "cust-1")
//
//    val event1 = CustomerCreated(id, command.name)
//    val event2 = CustomerActivated("a good reason", Instant.now())
//
//    val eventsList = asList(event1, event2)
//
//    val asJson = listOfEventsToJson(eventsList)
//
//    assertThat(listOfEventsFromJson(asJson)).isEqualTo(eventsList)
//
//  }

}
